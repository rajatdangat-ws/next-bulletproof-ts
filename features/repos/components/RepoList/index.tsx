/**
 *
 * RepoList
 *
 */

import React from "react";
import get from "lodash/get";
import { Skeleton } from "antd";
import { useRouter } from "next/router";
import { T, CustomCard, If } from "@common";
import { IResponse, ResponseItem } from "@features/repos/api/getRecommendations";

interface RepoListProps {
  reposData?: IResponse;
  loading: boolean;
  repoName: string;
}

const RepoList: React.FC<RepoListProps> = props => {
  const { reposData, loading, repoName } = props;
  const router = useRouter();

  const items: ResponseItem[] = get(reposData, "items", []);
  const totalCount: number = get(reposData, "totalCount", 0);
  const BlockText = props => <T display="block" {...props} />;

  return (
    <If condition={items.length !== 0 || loading}>
      <CustomCard data-testid="repo-list">
        <Skeleton loading={loading} active>
          {repoName && <BlockText id="search_query" values={{ repoName }} />}
          {totalCount !== 0 && <BlockText id="matching_repos" values={{ totalCount }} />}
          {items.map((item, index: number) => (
            <CustomCard
              key={index}
              onClick={() => router.push(`/info/${item?.name}?owner=${item?.owner.login}`)}
            >
              {item.name}
              <BlockText id="repository_name" values={{ name: item.name }} />
              <BlockText id="repository_full_name" values={{ fullName: item.fullName }} />
              <BlockText id="repository_stars" values={{ stars: item.stargazersCount }} />
            </CustomCard>
          ))}
        </Skeleton>
      </CustomCard>
    </If>
  );
};

export default RepoList;
