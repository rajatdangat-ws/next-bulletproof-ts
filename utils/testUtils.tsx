import { render, RenderOptions } from "@testing-library/react";
import React from "react";
import { IntlProvider } from "react-intl";
import { Provider as ReduxProvider } from "react-redux";
import { store } from "../store";
import messages from "../translations/en.json";

/**
 * Renders the passed in components or tree with all the providers.
 * To configure just add more providers and use the function in tests to render with providers
 * @param children, all the components in the composition
 * @returns React.FC, renders the DOM with all the providers
 */
export const WithAllProviders: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  return (
    <ReduxProvider store={store}>
      <IntlProvider messages={messages} locale="en" defaultLocale="en">
        {children}
      </IntlProvider>
    </ReduxProvider>
  );
};

/**
 * Generic render function that renders all the listed components with all the providers
 * @param ui The ReactElement (ie. Component) to render
 * @param options extra render options
 * @returns returns the RenderResult with all the things that react-testing library provides
 */
export const renderForTesting = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: WithAllProviders, ...options });

/**
 * Stops execution in tests for a specified milliseconds
 * @param ms time in milliseconds, how long to stop
 * @returns returns a new Promise to hang
 */
export const timeout = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));

/**
 * Generate an API Response that can be used to mock API Responses
 * @param ok boolean specifing will the response resolve or reject
 * @param data the data that is to be sent back
 * @returns APIResponse consisting ok and data
 */
export function apiResponseGenerator<T>(ok: boolean, data: T) {
  return {
    ok,
    data,
  };
}

export * from "@testing-library/react";
export { renderForTesting as render };
