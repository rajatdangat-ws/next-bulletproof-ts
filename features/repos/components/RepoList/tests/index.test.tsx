/**
 *
 * Tests for RepoList
 *
 */

import React from "react";
import { render } from "@utils/testUtils";
import RepoList from "../index";

describe("<RepoList />", () => {
  const repoListProps = {
    loading: true,
    repoName: "Nextjs Bulletproof Template",
    reposData: {
      totalCount: 1,
      incompleteResults: false,
      items: [
        {
          name: "Nextjs Bulletproof",
          fullName: "Nextjs Bulletproof Template",
          stargazersCount: 100,
          owner: { login: "Wednesday Solutions" },
        },
      ],
    },
  };

  it("should render and match the snapshot", () => {
    const { baseElement } = render(<RepoList {...repoListProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it("should contain 1 RepoList component", () => {
    const { getAllByTestId } = render(<RepoList {...repoListProps} />);
    expect(getAllByTestId("repo-list").length).toBe(1);
  });
});
