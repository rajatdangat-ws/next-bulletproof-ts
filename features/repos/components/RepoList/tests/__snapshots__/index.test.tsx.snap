// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RepoList /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered styled__CustomCard-sc-15gkald-1 kmxBSK"
      data-testid="repo-list"
    >
      <div
        class="ant-card-body"
      >
        <div
          class="ant-skeleton ant-skeleton-active"
        >
          <div
            class="ant-skeleton-content"
          >
            <h3
              class="ant-skeleton-title"
              style="width: 38%;"
            />
            <ul
              class="ant-skeleton-paragraph"
            >
              <li />
              <li />
              <li
                style="width: 61%;"
              />
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
